Network/Socket/ByteString/IO.hsc

```diff
@@ -269,22 +270,23 @@ remainingChunks i (x:xs)
 -- IOVec made from @cs@ and the number of pointers (@length cs@).
 -- /Unix only/.
 withIOVecfromBS :: [ByteString] -> ((Ptr IOVec, Int) -> IO a) -> IO a
-withIOVecfromBS cs f = do
-    bufsizs <- mapM getBufsiz cs
-    withIOVec bufsizs f
+withIOVecfromBS cs f = withBufSizs cs $ \bufsizs -> withIOVec bufsizs f
 #else
 -- | @withWSABuffromBS cs f@ executes the computation @f@, passing as argument a pair
 -- consisting of a pointer to a temporarily allocated array of pointers to
 -- WSABuf made from @cs@ and the number of pointers (@length cs@).
 -- /Windows only/.
 withWSABuffromBS :: [ByteString] -> ((Ptr WSABuf, Int) -> IO a) -> IO a
-withWSABuffromBS cs f = do
-    bufsizs <- mapM getBufsiz cs
-    withWSABuf bufsizs f
+withWSABuffromBS cs f = withBufSizs cs $ \bufsizs -> withWSABuf bufsizs f
 #endif
 
-getBufsiz :: ByteString -> IO (Ptr Word8, Int)
-getBufsiz (PS fptr off len) = withForeignPtr fptr $ \ptr -> return (ptr `plusPtr` off, len)
+withBufSizs :: [ByteString] -> ([(Ptr Word8, Int)] -> IO a) -> IO a
+withBufSizs bss0 f = loop bss0 id
+  where
+    loop []                    !build = f $ build []
+    loop (PS fptr off len:bss) !build = withForeignPtr fptr $ \ptr -> do
+        let !ptr' = ptr `plusPtr` off
+        loop bss (build . ((ptr',len) :))
```

Comment:
Here you've chosen to traverse the `ByteString` list just once via stack recursion through `withForeignPtr`, rather than build a list of `ForeignPtrs` to touch at the end.  I guess that makes sense.  The other option is to traverse the bytestring list twice, with just simple tail-recursive `withForeignPtr p1 $ \_ -> (withForeignPtr p2 $ \_ -> ... (WithForeignPtr pN $ \_ -> return result))))` at the end.  Not sure which is better...  The ByteString list should generally be quite short, so it probably doesn't matter much...

Reply:
I don't understand the other option.

Reply:
With `ForeignPtr.Unsafe` you can use `unsafeForeignPtrToPtr` get the list of pointers do the IO action as before, and then use the the `ByteString` list again to touch all the pointers:
```haskell
touchByteStrings ((PS f _ _):bss) ret1 =
    let ret2 = withForeignPtr f $ \_ -> return ret1
     in touchByteStrings bss ret2
touchByteStrings _ ret = ret
```
This allows you to write:
```
   bufs = map (\(PS f off len) -> (unsafeForeignPtrToPtr f `plusPtr` off, len)) bss
   ret <- f bufs
   touchByteStrings bss ret
```
without a stack depth equal to the `ByteString` count (admittedly never large).  But this requires the unsafe interface, and walks the `ByteString` list twice breaking single-pass lazy evaluation.  But we still need to allocate an `O(n)` iovec, so this is not a constant-space interface.

Reply:
I understand. Thank you. But let's stay with the current code. If we will face problems, I will switch to this approach.